syntax = "proto3";

package user;

// Service for user profile management
service UserService {
  rpc RegisterUser (RegisterUserRequest) returns (RegisterUserResponse);
  rpc LoginUser (LoginUserRequest) returns (LoginUserResponse);
  rpc GetUserProfile (GetUserProfileRequest) returns (GetUserProfileResponse);
  rpc UpdateUserProfile (UpdateUserProfileRequest) returns (UpdateUserProfileResponse);
  rpc DeleteUser (DeleteUserRequest) returns (DeleteUserResponse);
  rpc GetUserRecommendations(GetUserRecommendationsRequest) returns (GetUserRecommendationsResponse);
  rpc UpdatePreferences(UpdatePreferencesRequest) returns (UpdatePreferencesResponse);
}

// Request and response for RegisterUser
message RegisterUserRequest {
  string email = 1;
  string password = 2;
  string name = 3;
  string preferences = 4; // JSON string
}

message RegisterUserResponse {
    string message = 1;
    bool success = 2;
    string recommendations = 3; // Recommendations for first-time users
}

// Request and response for LoginUser
message LoginUserRequest {
  string email = 1;
  string password = 2;
}

message LoginUserResponse {
    string message = 1;
    bool success = 2;
    string token = 3; // JWT token for authentication
    int32 user_id = 4; // Add user_id to the response
    string recommendations = 5; // Optional field for returning recommendation
    string preferences = 6;
    string name = 7;
}

// Request and response for GetUserProfile
message GetUserProfileRequest {
  int32 user_id = 1;
}

message GetUserProfileResponse {
  int32 id = 1;
  string email = 2;
  string name = 3;
  string preferences = 4; // JSON string
}

// Request and response for UpdateUserProfile
message UpdateUserProfileRequest {
  int32 user_id = 1;
  string name = 2;
  string preferences = 3; // JSON string
}

message UpdateUserProfileResponse {
  string message = 1;
  bool success = 2;
}

// Request and response for DeleteUser
message DeleteUserRequest {
  int32 user_id = 1;
}

message DeleteUserResponse {
  string message = 1;
  bool success = 2;
}

message GetUserRecommendationsRequest {
    string token = 1; // JWT token for authentication
}

message Recommendation {
    string name = 1;
    string category = 2;
    float rating = 3;
    int32 review_count = 4;
    string city = 5;
    string address = 6;
    string phone = 7;
    string price = 8;
    string image_url = 9;
    string url = 10;
}

message GetUserRecommendationsResponse {
    repeated Recommendation recommendations = 1;
}

message UpdatePreferencesRequest {
    int32 user_id = 1;
    repeated string category = 2;
    string city = 3;
}

message UpdatePreferencesResponse {
    string message = 1;
    bool success = 2;
    string recommendations = 3; // New recommendations based on updated preferences
}